src/scheduler.cc

```diff
@@ -762,57 +759,56 @@ void SchedulerService::decrement_ref_count(const std::vector<ObjRef> &objrefs) {
 void SchedulerService::upstream_objrefs(ObjRef objref, std::vector<ObjRef> &objrefs) {
   // upstream_objrefs assumes that the lock reverse_target_objrefs_lock_ has been acquired
   objrefs.push_back(objref);
-  for (int i = 0; i < reverse_target_objrefs_[objref].size(); ++i) {
-    upstream_objrefs(reverse_target_objrefs_[objref][i], objrefs);
+  for (int i = 0; i < (*reverse_target_objrefs_.unsafe_get())[objref].size(); ++i) {
+    upstream_objrefs((*reverse_target_objrefs_.unsafe_get())[objref][i], objrefs);
   }
 }
 
 void SchedulerService::get_equivalent_objrefs(ObjRef objref, std::vector<ObjRef> &equivalent_objrefs) {
-  std::lock_guard<std::mutex> target_objrefs_lock(target_objrefs_lock_);
+  auto target_objrefs_ptr = target_objrefs_.get();
   ObjRef downstream_objref = objref;
-  while (target_objrefs_[downstream_objref] != downstream_objref && target_objrefs_[downstream_objref] != UNITIALIZED_ALIAS) {
+  while ((*target_objrefs_ptr)[downstream_objref] != downstream_objref && (*target_objrefs_ptr)[downstream_objref] != UNITIALIZED_ALIAS) {
     RAY_LOG(RAY_ALIAS, "Looping in get_equivalent_objrefs");
-    downstream_objref = target_objrefs_[downstream_objref];
+    downstream_objref = (*target_objrefs_ptr)[downstream_objref];
   }
-  std::lock_guard<std::mutex> reverse_target_objrefs_lock(reverse_target_objrefs_lock_);
+  auto reverse_target_objrefs_ptr = reverse_target_objrefs_.get();
   upstream_objrefs(downstream_objref, equivalent_objrefs);
 }
 
-// This method defines the order in which locks should be acquired.
-void SchedulerService::do_on_locks(bool lock) {
```

Comment:
Instead of removing `do_on_locks`, I would say just keep it and add

		typedef Mutex mutex_type;
		mutex_type& mutex() { return mutex_; }

to `Synchronized` and get the addresses of the mutexes via `&field_.mutex()`.